### PR TITLE
Enable trimming for Http.Abstractions

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs [hunk 1]
```diff
@@ -117,7 +124,7 @@ public static class UseMiddlewareExtensions
 
     private static IApplicationBuilder UseMiddlewareInterface(
         IApplicationBuilder app,
-        [DynamicallyAccessedMembers(MiddlewareAccessibility)] Type middlewareType)
```
Comment (davidfowl) [inline src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs:120]: What's up with this?

File: src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs [hunk 2]
```diff
@@ -61,46 +61,53 @@ public static class UseMiddlewareExtensions
...
+            throw new InvalidOperationException(Resources.FormatException_UseMiddlewareNonTaskReturnType(InvokeMethodName, InvokeAsyncMethodName, nameof(Task)));
+        }
+
+        var parameters = invokeMethod.GetParameters();
+        if (parameters.Length == 0 || parameters[0].ParameterType != typeof(HttpContext))
+        {
+            throw new InvalidOperationException(Resources.FormatException_UseMiddlewareNoParameters(InvokeMethodName, InvokeAsyncMethodName, nameof(HttpContext)));
+        }
+
+        var state = new InvokeMiddlewareState(middleware);
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs:95]: I briefly looked at this in the previous trimming PR. I think the issue is the DynamicallyAccessedMembers metadata on the middleware type doesn't flow through the closure. Is that right @pranavkm? Could this be a bug?

[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs:95]: I read further into the PR: https://github.com/dotnet/linker/issues/1981 😄

File: src/Shared/PropertyHelper/PropertyHelper.cs [hunk 3]
```diff
@@ -3,22 +3,20 @@
...
 
 [assembly: MetadataUpdateHandler(typeof(Microsoft.Extensions.Internal.PropertyHelper))]
 
 namespace Microsoft.Extensions.Internal;
 
-internal class PropertyHelper
+internal sealed class PropertyHelper
 {
     private const BindingFlags DeclaredOnlyLookup = BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static | BindingFlags.DeclaredOnly;
+    private const BindingFlags Everything = BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Static | BindingFlags.Instance;
```
Comment (pranavkm) [inline src/Shared/PropertyHelper/PropertyHelper.cs:19]: We were previously calling [`GetRuntimeProperties`](https://source.dot.net/#System.Private.CoreLib/RuntimeReflectionExtensions.cs,27) inside of a Linq expression which made the trimmer sad. I switched it to call the underlying APIs in a iterative loop.

File: src/Shared/PropertyHelper/PropertyHelper.cs [hunk 4]
```diff
@@ -492,36 +473,63 @@ private static PropertyHelper CreateInstance(PropertyInfo property)
...
+    /// </returns>
+    // There isn't a way to represent trimmability requirements since for type since we unwrap nullable types.
+    [RequiresUnreferencedCode("This API is not trim safe.")]
+    public static PropertyHelper[] GetProperties(
+        Type type,
+        ConcurrentDictionary<Type, PropertyHelper[]>? cache)
     {
         // Unwrap nullable types. This means Nullable<T>.Value and Nullable<T>.HasValue will not be
         // part of the sequence of properties returned by this method.
         type = Nullable.GetUnderlyingType(type) ?? type;
```
Comment (pranavkm) [inline src/Shared/PropertyHelper/PropertyHelper.cs:496]: Apparently this makes the whole code path not trimmer friendly. The DynamicallyAccessedMembers annotations are on `type`, but our attempt to unwrap nullable types messes with it because the returned value does not have any annotations. I wonder if I'm holding it wrong or if the trimmer needs to special case this API.

@vitek-karas / @eerhardt?

File: src/Tools/LinkabilityChecker/LinkabilityChecker.csproj [hunk 5]
```diff
@@ -22,6 +22,12 @@
       <!-- Link all assemblies -->
       <ILLinkArgs>$(ILLinkArgs) --action link</ILLinkArgs>
       <ILLinkArgs Condition="'$(GenerateLinkerWarningSuppressions)' == 'true'">$(ILLinkArgs) --generate-warning-suppressions xml</ILLinkArgs>
+      <!-- suppress warnings with the following codes:
+           IL2026: Calling A which has B can break functionality when trimming application code. The target method might be removed.
```
Comment (pranavkm) [inline src/Tools/LinkabilityChecker/LinkabilityChecker.csproj:26]: I guess the trimmer produces warnings if you have `RequiresUnreferencedCode` in your code which we do. The runtime also appears to suppress a bunch of these warnings, maybe these are primarily for applications and not for libraries?

File: src/Shared/PropertyHelper/PropertyHelper.cs (hunk 4 above)
Comment (vitek-karas) [reply src/Shared/PropertyHelper/PropertyHelper.cs:496]: This is a tricky one. From the trimmer (IL) perspective `Nullable<TestType>` is a completely different type from `TestType` and calling `GetProperties` on one has no implications to the other. So applying `PublicProperties` annotation to the `Nullable<TestType>` doesn't mean that it will preserve properties on `TestType`.

In order to make this work the way this code would need we would have to:
* Special case `Nullable<T>` and when applying annotations to it, we would have to apply them to both `Nullable<>` and `T`.
* Special case `Nullable<>.GetUnderlyingType` and propagate the annotation from the input `type` to the return value of this method.

It's definitely doable.

Technically this is a bigger problem - one can imagine that similar logic could be applied to for example `Lazy<T>`, or the Roslyn's `Optional<T>` and so on. But I don't know if it's worth introducing a new kind of annotation to be able to apply this to any generic (something like `class Lazy<[PropagateAnnotationToThis] T>`).

I created https://github.com/dotnet/linker/issues/2662 to track this.
Please let us know how important is this - meanwhile I'll initiate a discussion on the linker team to come up with the best solution to the problem.

File: src/Shared/PropertyHelper/PropertyHelper.cs [hunk 6]
```diff
@@ -42,11 +40,11 @@ internal class PropertyHelper
     private static readonly ConcurrentDictionary<Type, PropertyHelper[]> PropertiesCache = new();
 
     private static readonly ConcurrentDictionary<Type, PropertyHelper[]> VisiblePropertiesCache = new();
-
     // We need to be able to check if a type is a 'ref struct' - but we need to be able to compile
     // for platforms where the attribute is not defined, like net46. So we can fetch the attribute
     // by late binding. If the attribute isn't defined, then we assume we won't encounter any
     // 'ref struct' types.
+    [UnconditionalSuppressMessage("Trimming", "IL2026:Members annotated with 'RequiresUnreferencedCodeAttribute' require dynamic access otherwise can break functionality when trimming application code")]
```
Comment (vitek-karas) [inline src/Shared/PropertyHelper/PropertyHelper.cs:47]: Why do you need to suppress a warning here. Linker should be able to figure this out - as in it should be able to find the attribute type in the app/closure and not warn on this at all.
If you have a repro where this warns, could you please send it over to me?

File: src/Tools/LinkabilityChecker/LinkabilityChecker.csproj (hunk 5 above)
Comment (vitek-karas) [reply src/Tools/LinkabilityChecker/LinkabilityChecker.csproj:26]: It should not. It should only produce a warning if you have code which is not annotated which calls into something which has RUC on it.

This is VERY likely hiding a trimming hole somewhere - please don't do this!

Runtime does suppress some of these warnings, but it does so selectively at the point where they occur and we spent a LOT of time justifying the suppression.

In general my advice would be - treat any suppression as a bug. It can be a bug in the linker (like the various workaround for lambdas and so on) or bug in the annotated code, but it's almost always a bug. The cases where it's not are rare and require deep understanding of the area and the trimmer behavior (which we typically don't want to take a dependency on anyway).

File: src/Http/Http.Abstractions/src/Microsoft.AspNetCore.Http.Abstractions.WarningSuppressions.xml [hunk 7]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8"?>
+<linker>
```
Comment (vitek-karas) [inline src/Http/Http.Abstractions/src/Microsoft.AspNetCore.Http.Abstractions.WarningSuppressions.xml:2]: All of these are effectively a bug.

Suppressions are dangerous and should fall into two buckets:
* Something which is truly not problematic and we've spent the effort to prove that it's not a problem. In this case it should be suppressed directly in the code using the `UnconditionalSuppressMessageAttribute` on the affected method - with a justification/comment of why we thing it's OK. These should be rare - most likely just workarounds for limitations in the trimmer.
* Special cases, where we want the warning to be shown to the user and we can't fully annotate an API - this is done in runtime in only a handfull of places and it's typically around special features which are partially handled by native code (and so it's invisible to the trimmer). This should be SUPER rare and in almost all cases hidden behind a feature switch.

Runtime uses the XML attributes to suppress warnings in two scenarios:
* Warnings which we didn't get to fixing yet - so they are truly bugs, but we wanted to baseline the warnings for the build.
* Warnings which we want to keep in the code (as mentioned above) intentionally (very rare) - again we want to baseline these for the build.

Notice that in all of these cases the suppressions are NOT present in the final assembly used by developers - they're only passed to the trimmer on a command line - to effectively baseline (Selectively suppress) warnings during our own builds.

File: src/Http/Http.Abstractions/src/Microsoft.AspNetCore.Http.Abstractions.csproj [hunk 8]
```diff
@@ -13,6 +13,8 @@ Microsoft.AspNetCore.Http.HttpResponse</Description>
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
     <PackageTags>aspnetcore</PackageTags>
     <IsPackable>false</IsPackable>
+    <Trimmable>true</Trimmable>
+    <EnableTrimAnalyzer>true</EnableTrimAnalyzer>
```
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/Http/Http.Abstractions/src/Microsoft.AspNetCore.Http.Abstractions.csproj:17]: It doesn't appear to be enabled. I tried enabling it globally as part of this PR, but there's some stuff in Blazor that needs to be updated first.

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/Http/Http.Abstractions/src/Microsoft.AspNetCore.Http.Abstractions.csproj:17]: That's odd. Here's the code in the SDK that enables it:

https://github.com/dotnet/sdk/blob/1b963d8e57363c53a5a9dd1153ef939eb03f3d94/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.Analyzers.targets#L66

It _should_ be enabled by just setting IsTrimmable.

cc @sbomer

[the remark this answers is not in the document]
Comment (sbomer) [reply src/Http/Http.Abstractions/src/Microsoft.AspNetCore.Http.Abstractions.csproj:17]: Is this a problem with `Trimmable` vs `IsTrimmable`?

File: src/Shared/PropertyHelper/PropertyHelper.cs (hunk 6 above)
Comment (pranavkm) [reply src/Shared/PropertyHelper/PropertyHelper.cs:47]: The analyzer warning me about it and offered this suppression as a codefix. We could look at it at the same time we look at figuring out the other suppressions you asked me not to configure.

File: src/Http/Http.Abstractions/src/Microsoft.AspNetCore.Http.Abstractions.csproj (hunk 8 above)
[the remark this answers is not in the document]
Comment (eerhardt) [reply src/Http/Http.Abstractions/src/Microsoft.AspNetCore.Http.Abstractions.csproj:17]: Yep - looks like it. My mistake - I completely missed that difference. Thanks @sbomer !

File: src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs [hunk 9]
```diff
@@ -22,7 +21,8 @@ public static class UseMiddlewareExtensions
     private static readonly MethodInfo GetServiceInfo = typeof(UseMiddlewareExtensions).GetMethod(nameof(GetService), BindingFlags.NonPublic | BindingFlags.Static)!;
 
     // We're going to keep all public constructors and public methods on middleware
-    private const DynamicallyAccessedMemberTypes MiddlewareAccessibility = DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.PublicMethods;
+    private const DynamicallyAccessedMemberTypes MiddlewareAccessibility =
+        DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods | DynamicallyAccessedMemberTypes.NonPublicConstructors;
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs:25]: I don't see https://github.com/dotnet/aspnetcore/blob/4888800d435b86e0d93c25724eefb33a84332537/src/Shared/ActivatorUtilities/ActivatorUtilities.cs using non-public constructors.

Also, looking at 

https://github.com/dotnet/aspnetcore/blob/4888800d435b86e0d93c25724eefb33a84332537/src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs#L61-L67

I only see Public methods being inspected.

File: src/Tools/LinkabilityChecker/LinkabilityChecker.csproj (hunk 5 above)
Comment (eerhardt) [reply src/Tools/LinkabilityChecker/LinkabilityChecker.csproj:26]: > See also: https://github.com/dotnet/runtime/blob/7671b5084a41944d7a0c1fb9674e354183ec722d/eng/illink.targets#L238-L255

Note that we do different types of trimming in the runtime repo, so it isn't useful to compare.

In the case that you pointed to, we only run a single assembly through the trimmer and trim out any dead code. This isn't our "trim analysis" step - that step is separate and runs on the whole shared framework together. In that step we don't globally suppress any warnings. That code is here:

https://github.com/dotnet/runtime/blob/7671b5084a41944d7a0c1fb9674e354183ec722d/src/libraries/sfx.proj#L47-L63

Comment (vitek-karas) [reply src/Tools/LinkabilityChecker/LinkabilityChecker.csproj:26]: What Eric says - that run of trimmer has basically analysis disabled because we don't want any of the warnings from it.

File: src/Shared/PropertyHelper/PropertyHelper.cs (hunk 6 above)
Comment (vitek-karas) [reply src/Shared/PropertyHelper/PropertyHelper.cs:47]: I would not trust the analyzer on this - just yet. There were lot of issues fixed in the analyzer recently for one thing. (Not saying that you should not pay attention and tells us about any discrepancies). The general rule we're trying hard to follow is:
- analyzer should not report warnings which the linker does not

So in short, if you see analyzer warning, but linker doesn't -> don't trust the analyzer, it's very probably wrong (and let us know so that we can fix it).

Comment (BrennanConroy) [reply src/Shared/PropertyHelper/PropertyHelper.cs:47]: Change to `private static readonly Type IsByRefLikeAttribute = typeof(System.Runtime.CompilerServices.IsByRefLikeAttribute);`

File: src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs:25]: Here's what I see if I remove the `PrivateConstructors` option:

```
Severity	Code	Description	Project	File	Line	Suppression State
Warning	IL2072	Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass5_0.<UseMiddleware>b__0(RequestDelegate): 'instanceType' argument does not satisfy 'DynamicallyAccessedMemberTypes.NonPublicConstructors' in call to 'Microsoft.Extensions.Internal.ActivatorUtilities.CreateInstance(IServiceProvider,Type,Object[])'. The return value of method 'Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.InvokeMiddlewareState.Middleware.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.	LinkabilityChecker (src\Tools\LinkabilityChecker\LinkabilityChecker)	D:\work\aspnetcore\src\Http\Http.Abstractions\src\Extensions\UseMiddlewareExtensions.cs	104	
```

File: src/Tools/LinkabilityChecker/LinkabilityChecker.csproj (hunk 5 above)
Comment (pranavkm) [reply src/Tools/LinkabilityChecker/LinkabilityChecker.csproj:26]: If I remove these warnings, I get a bunch of warnings I wasn't sure how to address. Here they are:

```C#
Severity	Code	Description	Project	File	Line	Suppression State
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'Microsoft.Extensions.Internal.PropertyHelper.<GetProperties>g__AddInterestingProperties|41_0(List<PropertyHelper>,Type)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'Microsoft.Extensions.Internal.PropertyHelper.GetProperties(Type)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'Microsoft.Extensions.Internal.PropertyHelper.GetVisibleProperties(Type)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'System.Delegate.CreateDelegate(Type,Type,String)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'System.Delegate.CreateDelegate(Type,Type,String,Boolean)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'System.Delegate.CreateDelegate(Type,Type,String,Boolean,Boolean)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'System.Delegate.Delegate(Type,String)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
Warning	IL2111	Microsoft.AspNetCore.Http.Abstractions: Method 'System.MulticastDelegate.MulticastDelegate(Type,String)' with parameters or return value with `DynamicallyAccessedMembersAttribute` is accessed via reflection. Trimmer can't guarantee availability of the requirements of the method.
```

Comment (pranavkm) [reply src/Tools/LinkabilityChecker/LinkabilityChecker.csproj:26]: I also get warnings for things which I thought have the right annotations, but the trimmer still warns:

```C#
Severity	Code	Description	Project	File	Line	Suppression State
Warning	IL2026	Microsoft.AspNetCore.Http.Abstractions: Using member 'Microsoft.AspNetCore.Routing.RouteValueDictionary.PropertyStorage.PropertyStorage(Object)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. This API is not trim safe.
Severity	Code	Description	Project	File	Line	Suppression State
Warning	IL2026	Microsoft.AspNetCore.Http.Abstractions: Using member 'Microsoft.AspNetCore.Routing.RouteValueDictionary.RouteValueDictionary(Object)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. This constructor may perform reflection on the specificed value which may be trimmed if not referenced directly. Consider using a different overload to avoid this issue.
```

Both of these methods have `RequiresUnreferencedCode`. Similarly a bunch of things in `PropertyHelpers` which are annotated with `RequiresUnreferencedCode` appear in the warning list.

Comment (vitek-karas) [reply src/Tools/LinkabilityChecker/LinkabilityChecker.csproj]: I looked into this a bit more.
The IL2111 on PropertyHelper is caused by the `MetadataUpdateHandler(typeof(PropertyHelper))`. The attribute has annotation `All` on the type - so it will force include everything on the class - and if there's something which has annotation it will warn about it. I don't know why it needs `All` - based on the description in the docs, it should be enough to have "all methods" (not that it would help here).
I can't think of a good way to fix this . You could suppress the warning on the entire type (PropertyHelper) with an explanation that it's because of the MetadataUpdateHandler and that it will only ever use the `ClearCache` method which doesn't have any annotation.
Alternatively, you could create a static subclass like `PropertyHelper.MetadataUpdateHandler` which would be used in the attribute and it would simply forward the call to the main class. That should not produce any warnings at all. Personally I would probably prefer that solution - it doesn't feel that bad.

I think the delegate related warnings are for the same reason - linker has some implicit rules about marking things on delegate types, so applying `All` to a type which declares a delegate would likely produce these warnings. If the attribute didn't have `All` this would probably go away (but the solution with a static subclass would also fix this).

Comment (vitek-karas) [reply src/Tools/LinkabilityChecker/LinkabilityChecker.csproj]: The tow IL2026 warnings are also very likely due to the `MetadataUpdateHandler` attribute, this time on `RouteValueDictionary`. Again, using a static subclass as the handler would solve the problem - but I would question the annotation on the attribute - it clearly should not need constructors or nested types.

Comment (vitek-karas) [reply src/Tools/LinkabilityChecker/LinkabilityChecker.csproj]: I created https://github.com/dotnet/linker/issues/2668 to provide a better location for these warnings which should make them easier to diagnose.

**Review by Tratcher — APPROVED** [review]
